app/models/population.rb

```diff
@@ -23,6 +23,14 @@ def free_spaces
     usable_capacity - unlock - bedwatch - overnights_in - moves_to.count + overnights_out + out_of_area_courts + discharges + moves_from.count
   end
 
+  def save_uniquely!
+    save!
+    self
+  rescue PG::UniqueViolation, ActiveRecord::RecordNotUnique
```

Comment:
Should we just add a validation that requires that the population is unique?

Reply:
Confused about why this can't just be handled in the same way as: https://github.com/ministryofjustice/hmpps-book-secure-move-api/blob/968a860c5b92df2b4f87eacaabe8e4cf7086abd3/app/models/move.rb#L99

Reply:
Obviously with a simpler approach to validating uniqueness

Reply:
I agree that a uniqueness validation would be good addition 👍 
but I think the database fallback catch is _also_ required, because the pre-emptive uniqueness validator doesn't always catch clashes (creating moves on different pods etc)

Reply:
I thought this would be controversial so it's great you all spotted it 😛 . Here's my thinking - the rails unique validations basically don't work as Abeer found out in previous work on the PER it's possible and has been observed for two requests to overlap and defeat Rails uniqueness validations (which essentially perform a query before an update). Without a database level constraint it's therefore possible to get duplicate rows; in the case of the population free spaces view this would lead to incorrect calculations and confusing duplicate data so we don't want that.

The database level constraint is therefore essential to ensure that no duplicates can be created. That means we have to catch database level exceptions and propagate them nicely back to the client (raising the same error message they would have seen with a rails level uniqueness check).

So that leads me to wonder what value the uniqueness check would be adding on top of that - it means performing an additional `SELECT` query before every update and insert of the population table, plus additional code that needs spec coverage, only to provide the same level of uniqueness protection we get solely from the database level constraint. I think it's basically superfluous in this circumstance which is why I left it out. It's less idiomatic than the usual approach but perhaps the usual approach is flawed... 🤯

Reply:
Lots of articles on this online but this captures the main points pretty well: https://makandracards.com/makandra/13901-understanding-race-conditions-with-duplicate-unique-keys-in-rails-mysql